File: app/src/main/java/com/storehop/app/data/dao/ItemStoreXrefDao.kt
package com.storehop.app.data.dao

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Transaction
import androidx.room.Upsert
import com.storehop.app.data.entity.ItemStoreXref
import kotlinx.coroutines.flow.Flow

/**
 * v0.7.0 access scope: queries filter by `householdId` (not `userId`).
 *
 * `userId` remains on each xref as creator/audit metadata — copied from
 * the parent item at insert time so cross-table ownership stays coherent
 * across session changes. `householdId` is what scopes who can see and
 * mutate the row. For single-member households both columns hold the
 * same value, so behaviour matches v0.6.x exactly.
 *
 * [setStoresForItem] takes both: `householdId` to scope the existing-row
 * lookup and to stamp on new rows, plus `userId` to carry the parent's
 * creator stamp through onto the freshly-inserted junction rows.
 */
@Dao
interface ItemStoreXrefDao {

    @Query(
        """
        SELECT storeId FROM item_store_xref
        WHERE itemId = :itemId AND deletedAt IS NULL
        """,
    )
    fun observeStoreIdsForItem(itemId: String): Flow<List<String>>

    @Query(
        """
        SELECT * FROM item_store_xref
        WHERE itemId = :itemId AND deletedAt IS NULL
        """,
    )
    suspend fun findForItem(itemId: String): List<ItemStoreXref>

    @Upsert
    suspend fun upsert(xref: ItemStoreXref)

    /**
     * Batch upsert for the pull side. Mappers stamp `pendingSync = false`.
     * Called inside [PullWriteDao]'s single transaction.
     */
    @Upsert
    suspend fun upsertFromCloud(rows: List<ItemStoreXref>)

    @Query(
        """
        UPDATE item_store_xref
        SET deletedAt = :now, updatedAt = :now, pendingSync = 1
        WHERE itemId = :itemId AND storeId = :storeId AND householdId = :householdId
        """,
    )
    suspend fun softDelete(householdId: String, itemId: String, storeId: String, now: Long)

    /**
     * Cascade-tombstone every live xref for an item. Used by the item soft-delete
     * flow so a deleted item doesn't leave xref orphans (the cross-store-sync
     * `ShoppingDao` query INNER JOINs through xrefs, so live xrefs against a
     * deleted item would surface ghost rows). Scoped by `householdId`.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET deletedAt = :now, updatedAt = :now, pendingSync = 1
        WHERE itemId = :itemId AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    suspend fun softDeleteForItem(householdId: String, itemId: String, now: Long)

    /**
     * Cascade-tombstone every live xref pointing at a store. Used by the store
     * soft-delete flow so items previously tagged to the deleted store don't
     * still appear "sold at <tombstoned store>" in any relation that joins
     * items to stores through this junction.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET deletedAt = :now, updatedAt = :now, pendingSync = 1
        WHERE storeId = :storeId AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    suspend fun softDeleteForStore(householdId: String, storeId: String, now: Long)

    /**
     * Restore the cascade-tombstoned xrefs for a store. Filters by the exact
     * `deletedAt` used during the cascade so a later, separate tombstoning of
     * a different row at the same time doesn't accidentally come back too.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET deletedAt = NULL, updatedAt = :now, pendingSync = 1
        WHERE storeId = :storeId AND householdId = :householdId AND deletedAt = :deletedAt
        """,
    )
    suspend fun restoreCascadeForStore(householdId: String, storeId: String, deletedAt: Long, now: Long)

    /** Inverse of [softDeleteForItem]; same `deletedAt`-filter pattern. */
    @Query(
        """
        UPDATE item_store_xref
        SET deletedAt = NULL, updatedAt = :now, pendingSync = 1
        WHERE itemId = :itemId AND householdId = :householdId AND deletedAt = :deletedAt
        """,
    )
    suspend fun restoreCascadeForItem(householdId: String, itemId: String, deletedAt: Long, now: Long)

    /**
     * Replace the set of stores an item is tagged to.
     * Tombstones any xref no longer in [storeIds] and upserts the new set.
     * `householdId` scopes who owns the row (the parent item's householdId);
     * `userId` is the parent's creator-stamp copied onto the new junction rows.
     *
     * v0.9: retained stores (already alive AND still selected) used to be left
     * completely untouched, so a store stranded at `isNeeded = 0` by a prior
     * cross-store purchase cascade stayed invisible on that store's list even
     * after the user re-saved the item with the store still ticked — the
     * "Aldi is selected but the item never appears" bug Mike reported. We now
     * converge `isNeeded` across the final store set to the item's current
     * global needed state: if the item is needed at ANY alive store it's
     * needed at all of them; if it's fully purchased everywhere we leave it
     * purchased (so editing a bought item's name doesn't resurrect it onto
     * every list).
     */
    @Transaction
    suspend fun setStoresForItem(
        itemId: String,
        storeIds: Set<String>,
        householdId: String,
        userId: String,
        now: Long,
    ) {
        val existing = findForItem(itemId)
        val existingIds = existing.map { it.storeId }.toSet()
        val toRemove = existingIds - storeIds
        val toAdd = storeIds - existingIds
        // Global needed state, evaluated from the alive xrefs BEFORE we mutate
        // anything. A newly-created item (no existing xrefs) counts as needed
        // so its first stores land on the list, matching the old
        // upsert-with-isNeeded=true default.
        val neededGlobally = existing.isEmpty() || existing.any { it.isNeeded }
        toRemove.forEach { softDelete(householdId, itemId, it, now) }
        toAdd.forEach { storeId ->
            upsert(
                ItemStoreXref(
                    itemId = itemId,
                    storeId = storeId,
                    userId = userId,
                    createdAt = now,
                    updatedAt = now,
                    deletedAt = null,
                    isNeeded = neededGlobally,
                    householdId = householdId,
                ),
            )
        }
        // Converge retained stores that are stranded at isNeeded = 0 while the
        // item is needed globally. Only touch the stranded ones so we don't
        // needlessly re-dirty already-correct rows for sync.
        if (neededGlobally) {
            existing.asSequence()
                .filter { it.storeId in storeIds && !it.isNeeded }
                .forEach { markNeededAtStore(householdId, itemId, it.storeId, now) }
        }
    }

    @androidx.room.Query("SELECT * FROM item_store_xref WHERE householdId = :householdId AND pendingSync = 1")
    fun observePendingPush(householdId: String): kotlinx.coroutines.flow.Flow<List<ItemStoreXref>>

    /** v0.7.1: row-count of pending pushes for the Force-sync-now UX. */
    @androidx.room.Query("SELECT COUNT(*) FROM item_store_xref WHERE householdId = :householdId AND pendingSync = 1")
    fun countPendingPush(householdId: String): kotlinx.coroutines.flow.Flow<Int>

    /**
     * v0.8.0.4: composite-PK snapshot of pendingSync = 1 xrefs in the
     * household. Read by [PullWriteDao] to filter cloud xrefs that
     * would otherwise resurrect local edits.
     */
    @androidx.room.Query(
        "SELECT itemId, storeId FROM item_store_xref WHERE householdId = :householdId AND pendingSync = 1"
    )
    suspend fun pendingPushKeys(householdId: String): List<XrefKey>

    @androidx.room.Query(
        """
        UPDATE item_store_xref SET pendingSync = 0
        WHERE itemId = :itemId AND storeId = :storeId AND householdId = :householdId
        """,
    )
    suspend fun markPushed(householdId: String, itemId: String, storeId: String)

    /**
     * Mark this single (itemId, storeId) xref purchased at this store only.
     * Used by [markNeededAtStore]'s mirror — the per-store state-correction
     * path. The Shop-at-Store check-off cascades via
     * [markPurchasedAcrossAllStores] instead, which writes the same flips to
     * every store the item is tagged to.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 0,
            lastPurchasedAt = :now,
            updatedAt = :now,
            pendingSync = 1
        WHERE itemId = :itemId AND storeId = :storeId AND householdId = :householdId
        """,
    )
    suspend fun markPurchasedAtStore(householdId: String, itemId: String, storeId: String, now: Long)

    /**
     * Cascade-mark every live xref for [itemId] as purchased: a single
     * shopping trip (one trip to one store) satisfies the need across every
     * store the item is tagged to. Used by the Shop-at-Store check-off so
     * that buying mozzarella at Lidl makes it disappear from the Aldi and
     * Pingo Doce lists too, matching how shopping-list apps usually behave.
     *
     * All affected rows share the same `lastPurchasedAt = :now`, which is
     * the precision marker [restorePurchaseAcrossAllStores] uses to undo a
     * cascade without touching unrelated history.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 0,
            lastPurchasedAt = :now,
            updatedAt = :now,
            pendingSync = 1
        WHERE itemId = :itemId AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    suspend fun markPurchasedAcrossAllStores(householdId: String, itemId: String, now: Long)

    /**
     * Restore the (itemId, storeId) row to needed at this store. Used to
     * undo a mis-tap in the Shop-at-Store screen. lastPurchasedAt is left
     * intact -- the prior purchase still happened in history.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 1,
            updatedAt = :now,
            pendingSync = 1
        WHERE itemId = :itemId AND storeId = :storeId AND householdId = :householdId
        """,
    )
    suspend fun markNeededAtStore(householdId: String, itemId: String, storeId: String, now: Long)

    /**
     * Inverse of [markPurchasedAcrossAllStores], filtered by exact
     * `lastPurchasedAt` so concurrent or older purchases of the same item
     * aren't accidentally restored. Restores the xrefs to `isNeeded = 1`
     * and clears `lastPurchasedAt`, since the snackbar-undo flow also
     * deletes the matching `PurchaseRecord` — it's "as if it never
     * happened," not a state correction.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 1,
            lastPurchasedAt = NULL,
            updatedAt = :now,
            pendingSync = 1
        WHERE itemId = :itemId AND householdId = :householdId
            AND lastPurchasedAt = :lastPurchasedAt
            AND isNeeded = 0
        """,
    )
    suspend fun restorePurchaseAcrossAllStores(
        householdId: String,
        itemId: String,
        lastPurchasedAt: Long,
        now: Long,
    )

    /**
     * Set every alive xref for [itemId] to needed. Used by the v0.6.1
     * Items-list "+" tap: when the user adds an item from the master list,
     * we mark it needed at every store it's tagged to. Cross-store cascade
     * (one trip clears the list everywhere) means the inverse "−" tap can
     * use the existing [markPurchasedAcrossAllStores] without writing a
     * PurchaseRecord — this is a list-state action, not a purchase event.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 1,
            lastPurchasedAt = NULL,
            updatedAt = :now,
            pendingSync = 1
        WHERE itemId = :itemId AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    suspend fun markNeededAcrossAllStores(householdId: String, itemId: String, now: Long)

    /**
     * Distinct item IDs that have at least one alive xref with `isNeeded = 1`
     * for the given household. Powers the v0.6.1 +/− toggle on the Items
     * list: the screen shows "−" when the item is on the list at any
     * tagged store, "+" otherwise.
     */
    @Query(
        """
        SELECT DISTINCT itemId FROM item_store_xref
        WHERE householdId = :householdId AND deletedAt IS NULL AND isNeeded = 1
        """,
    )
    fun observeNeededItemIds(householdId: String): Flow<List<String>>

    /**
     * v0.9 repair pass (Force Sync now): converge every item to global-needed.
     * For any item that is needed at ≥1 alive store, flip every OTHER alive
     * store for that item to needed too. Heals rows stranded at `isNeeded = 0`
     * by the pre-v0.9 asymmetric un-check (which only restored the store you
     * were standing in). Idempotent — items needed nowhere are left purchased,
     * and already-converged items match nothing. Flipped rows are flagged
     * `pendingSync = 1` so the same Force-Sync flush pushes them up.
     *
     * Returns the number of rows repaired.
     */
    @Query(
        """
        UPDATE item_store_xref
        SET isNeeded = 1, updatedAt = :now, pendingSync = 1
        WHERE householdId = :householdId
          AND deletedAt IS NULL
          AND isNeeded = 0
          AND itemId IN (
            SELECT itemId FROM item_store_xref
            WHERE householdId = :householdId AND deletedAt IS NULL AND isNeeded = 1
          )
        """,
    )
    suspend fun repairStrandedNeededLinks(householdId: String, now: Long): Int
}

/**
 * v0.8.0.4: composite primary-key tuple for [ItemStoreXrefDao.pendingPushKeys].
 * Used by [PullWriteDao] to filter cloud xrefs against a Set of local
 * pending keys before upsert — preserves the user's most recent intent
 * when a pull races a push.
 */
data class XrefKey(val itemId: String, val storeId: String)

File: ios/Storehop/Data/DAO/ItemStoreXrefDao.swift
import Foundation
import GRDB

/// Owns per-store shopping state since migration v4→v5. Two operations are
/// load-bearing:
///
/// 1. `setStoresForItem` — diff/upsert that the form save flow uses.
/// 2. `markPurchasedAcrossAllStores` — v0.5.1 cross-store cascade. One trip
///    to Lidl marks mozzarella purchased at Aldi and Pingo Doce too.
///
/// v0.7.0 access scope: queries filter by `householdId` (not `userId`).
/// `userId` remains on each xref as creator/audit metadata — copied from
/// the parent item at insert time so cross-table ownership stays coherent
/// across session changes. `householdId` is what scopes who can see and
/// mutate the row. For single-member households both columns hold the
/// same value, so behaviour matches v0.6.x exactly.
///
/// `setStoresForItem` takes both: `householdId` to scope the existing-row
/// lookup and to stamp on new rows, plus `userId` to carry the parent's
/// creator stamp through onto the freshly-inserted junction rows.
struct ItemStoreXrefDao: Sendable {
    let writer: any DatabaseWriter

    // MARK: - Reactive

    func observeStoreIdsForItem(itemId: String) -> AsyncValueObservation<[String]> {
        ValueObservation
            .tracking { db in
                try String.fetchAll(db, sql: """
                    SELECT storeId FROM item_store_xref
                    WHERE itemId = ? AND deletedAt IS NULL
                    """, arguments: [itemId])
            }
            .values(in: writer)
    }

    /// v0.6.1: distinct item IDs that have at least one alive xref with
    /// `isNeeded = 1` for the given household. Powers the +/- toggle on the
    /// Items list -- "−" when the item is on the list at any tagged store,
    /// "+" otherwise.
    func observeNeededItemIds(householdId: String) -> AsyncValueObservation<[String]> {
        ValueObservation
            .tracking { db in
                try String.fetchAll(db, sql: """
                    SELECT DISTINCT itemId FROM item_store_xref
                    WHERE householdId = ? AND deletedAt IS NULL AND isNeeded = 1
                    """, arguments: [householdId])
            }
            .values(in: writer)
    }

    /// v0.7.1: row-count of pending pushes for the Force-sync-now UX.
    func countPendingPush(householdId: String) -> AsyncValueObservation<Int> {
        ValueObservation
            .tracking { db in
                try Int.fetchOne(db, sql: """
                    SELECT COUNT(*) FROM item_store_xref
                    WHERE householdId = ? AND pendingSync = 1
                    """, arguments: [householdId]) ?? 0
            }
            .values(in: writer)
    }

    func observePendingPush(householdId: String) -> AsyncValueObservation<[ItemStoreXref]> {
        ValueObservation
            .tracking { db in
                try ItemStoreXref.fetchAll(db, sql: """
                    SELECT * FROM item_store_xref
                    WHERE householdId = ? AND pendingSync = 1
                    """, arguments: [householdId])
            }
            .values(in: writer)
    }

    // MARK: - Snapshot

    func findForItem(itemId: String) async throws -> [ItemStoreXref] {
        try await writer.read { db in try Self.findForItem(on: db, itemId: itemId) }
    }

    static func findForItem(on db: Database, itemId: String) throws -> [ItemStoreXref] {
        try ItemStoreXref.fetchAll(db, sql: """
            SELECT * FROM item_store_xref
            WHERE itemId = ? AND deletedAt IS NULL
            """, arguments: [itemId])
    }

    // MARK: - Writes

    func upsert(_ xref: ItemStoreXref) async throws {
        try await writer.write { db in
            var copy = xref
            try copy.upsert(db)
        }
    }

    func upsertFromCloud(_ rows: [ItemStoreXref], on db: Database) throws {
        for row in rows {
            var copy = row
            try copy.upsert(db)
        }
    }

    func softDelete(householdId: String, itemId: String, storeId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.softDelete(on: db, householdId: householdId, itemId: itemId, storeId: storeId, now: now)
        }
    }

    static func softDelete(on db: Database, householdId: String, itemId: String, storeId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET deletedAt = ?, updatedAt = ?, pendingSync = 1
            WHERE itemId = ? AND storeId = ? AND householdId = ?
            """, arguments: [now, now, itemId, storeId, householdId])
    }

    func softDeleteForItem(householdId: String, itemId: String, now: Int64) async throws {
        try await writer.write { db in try Self.softDeleteForItem(on: db, householdId: householdId, itemId: itemId, now: now) }
    }

    static func softDeleteForItem(on db: Database, householdId: String, itemId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET deletedAt = ?, updatedAt = ?, pendingSync = 1
            WHERE itemId = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [now, now, itemId, householdId])
    }

    func softDeleteForStore(householdId: String, storeId: String, now: Int64) async throws {
        try await writer.write { db in try Self.softDeleteForStore(on: db, householdId: householdId, storeId: storeId, now: now) }
    }

    static func softDeleteForStore(on db: Database, householdId: String, storeId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET deletedAt = ?, updatedAt = ?, pendingSync = 1
            WHERE storeId = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [now, now, storeId, householdId])
    }

    func restoreCascadeForStore(householdId: String, storeId: String, deletedAt: Int64, now: Int64) async throws {
        try await writer.write { db in
            try Self.restoreCascadeForStore(on: db, householdId: householdId, storeId: storeId, deletedAt: deletedAt, now: now)
        }
    }

    static func restoreCascadeForStore(on db: Database, householdId: String, storeId: String, deletedAt: Int64, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET deletedAt = NULL, updatedAt = ?, pendingSync = 1
            WHERE storeId = ? AND householdId = ? AND deletedAt = ?
            """, arguments: [now, storeId, householdId, deletedAt])
    }

    func restoreCascadeForItem(householdId: String, itemId: String, deletedAt: Int64, now: Int64) async throws {
        try await writer.write { db in
            try Self.restoreCascadeForItem(on: db, householdId: householdId, itemId: itemId, deletedAt: deletedAt, now: now)
        }
    }

    static func restoreCascadeForItem(on db: Database, householdId: String, itemId: String, deletedAt: Int64, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET deletedAt = NULL, updatedAt = ?, pendingSync = 1
            WHERE itemId = ? AND householdId = ? AND deletedAt = ?
            """, arguments: [now, itemId, householdId, deletedAt])
    }

    /// Replace the set of stores an item is tagged to. Tombstones any xref
    /// no longer in `storeIds` and upserts the new set.
    /// `householdId` scopes who owns the row (the parent item's householdId);
    /// `userId` is the parent's creator-stamp copied onto the new junction rows.
    func setStoresForItem(itemId: String, storeIds: Set<String>, householdId: String, userId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.setStoresForItem(on: db, itemId: itemId, storeIds: storeIds, householdId: householdId, userId: userId, now: now)
        }
    }

    static func setStoresForItem(on db: Database, itemId: String, storeIds: Set<String>, householdId: String, userId: String, now: Int64) throws {
        let existing = try findForItem(on: db, itemId: itemId)
        let existingIds = Set(existing.map(\.storeId))
        let toRemove = existingIds.subtracting(storeIds)
        let toAdd = storeIds.subtracting(existingIds)

        for storeId in toRemove {
            try softDelete(on: db, householdId: householdId, itemId: itemId, storeId: storeId, now: now)
        }
        for storeId in toAdd {
            var xref = ItemStoreXref(
                itemId: itemId,
                storeId: storeId,
                userId: userId,
                createdAt: now,
                updatedAt: now,
                deletedAt: nil,
                pendingSync: true,
                isNeeded: true,
                lastPurchasedAt: nil,
                householdId: householdId
            )
            try xref.upsert(db)
        }
    }

    func markPushed(householdId: String, itemId: String, storeId: String) async throws {
        try await writer.write { db in
            try db.execute(sql: """
                UPDATE item_store_xref SET pendingSync = 0
                WHERE itemId = ? AND storeId = ? AND householdId = ?
                """, arguments: [itemId, storeId, householdId])
        }
    }

    // MARK: - Per-store check-off

    func markPurchasedAtStore(householdId: String, itemId: String, storeId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.markPurchasedAtStore(on: db, householdId: householdId, itemId: itemId, storeId: storeId, now: now)
        }
    }

    static func markPurchasedAtStore(on db: Database, householdId: String, itemId: String, storeId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET isNeeded = 0,
                lastPurchasedAt = ?,
                updatedAt = ?,
                pendingSync = 1
            WHERE itemId = ? AND storeId = ? AND householdId = ?
            """, arguments: [now, now, itemId, storeId, householdId])
    }

    func markPurchasedAcrossAllStores(householdId: String, itemId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.markPurchasedAcrossAllStores(on: db, householdId: householdId, itemId: itemId, now: now)
        }
    }

    static func markPurchasedAcrossAllStores(on db: Database, householdId: String, itemId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET isNeeded = 0,
                lastPurchasedAt = ?,
                updatedAt = ?,
                pendingSync = 1
            WHERE itemId = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [now, now, itemId, householdId])
    }

    /// v0.6.1: inverse of [markPurchasedAcrossAllStores]. Sets every alive
    /// xref for [itemId] to `isNeeded = 1, lastPurchasedAt = NULL`. Used by
    /// the "+" tap on the Items list to add an item to every tagged store
    /// without writing a PurchaseRecord -- the user is on the master list,
    /// not at a store.
    func markNeededAcrossAllStores(householdId: String, itemId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.markNeededAcrossAllStores(on: db, householdId: householdId, itemId: itemId, now: now)
        }
    }

    static func markNeededAcrossAllStores(on db: Database, householdId: String, itemId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET isNeeded = 1,
                lastPurchasedAt = NULL,
                updatedAt = ?,
                pendingSync = 1
            WHERE itemId = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [now, itemId, householdId])
    }

    func markNeededAtStore(householdId: String, itemId: String, storeId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.markNeededAtStore(on: db, householdId: householdId, itemId: itemId, storeId: storeId, now: now)
        }
    }

    static func markNeededAtStore(on db: Database, householdId: String, itemId: String, storeId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET isNeeded = 1,
                updatedAt = ?,
                pendingSync = 1
            WHERE itemId = ? AND storeId = ? AND householdId = ?
            """, arguments: [now, itemId, storeId, householdId])
    }

    func restorePurchaseAcrossAllStores(householdId: String, itemId: String, lastPurchasedAt: Int64, now: Int64) async throws {
        try await writer.write { db in
            try Self.restorePurchaseAcrossAllStores(on: db, householdId: householdId, itemId: itemId, lastPurchasedAt: lastPurchasedAt, now: now)
        }
    }

    static func restorePurchaseAcrossAllStores(on db: Database, householdId: String, itemId: String, lastPurchasedAt: Int64, now: Int64) throws {
        try db.execute(sql: """
            UPDATE item_store_xref
            SET isNeeded = 1,
                lastPurchasedAt = NULL,
                updatedAt = ?,
                pendingSync = 1
            WHERE itemId = ? AND householdId = ?
              AND lastPurchasedAt = ?
              AND isNeeded = 0
            """, arguments: [now, itemId, householdId, lastPurchasedAt])
    }
}
